semeio/jobs/ots/ots.py

```diff
@@ -0,0 +1,477 @@
+
+        xstart = np.min(x)
+        ystart = np.min(y)
+
+        if nx < 2 or ny < 2:
+            raise RuntimeError("Cannot create IRAP surface if nx or ny is <2")
+
+        xinc = (np.max(x) - xstart) / (nx - 1)
+        yinc = (np.max(y) - ystart) / (ny - 1)
+
+        surf = Surface(
+            nx=nx, ny=ny, xinc=xinc, yinc=yinc, xstart=xstart, ystart=ystart, angle=0
+        )
+
+        irap_x = np.empty(nx * ny)
+        irap_y = np.array(irap_x)
+
+        for i in range(len(surf)):
+            irap_x[i], irap_y[i] = surf.getXY(i)
+
+        # Interpolate vel grid to irap grid, should be the same apart from ordering
+        z = np.nan_to_num(z)
+        ip = CloughTocher2DInterpolator((x, y), z, fill_value=0)
+        irap_z = ip(irap_x, irap_y)
+
+        for i in range(len(surf)):
+            surf[i] = irap_z[i]
+
+        return surf
+
+    def add_survey(self, name, date):
+        """The add_survey() method will register a survey at a specific date.
+
+        The name argument should be a unique string, this will later
+        be used when evaluating the elastic strain in the
+        overburden. The date should be python datetime.date()
+        instance, this date should be present as a report step in the
+        restart file - otherwise an exception will be raised.
+        """
+        restart_view = self._rst_file.restartView(sim_time=date)
+        self.subsidence.add_survey_PRESSURE(name, restart_view)
+
+        self._restart_views[name] = restart_view
+
+        return restart_view
+
+    @staticmethod
+    def _divide_negative_shift(ts, div_val=5.0):
+        for i in range(len(ts)):
+            if ts[i] < 0:
+                ts[i] /= div_val
+
+    def geertsma_ts_rporv(self, vintage_pairs):
+        """
+        Calculates TS without using velocity. Fast.
+        Velocity is only used to get the surface on the velocity grid.
+        Uses change in porevolume from Eclipse (RPORV in .UNRST) as input to
+        Geertsma model.
+
+        :param vintage_pairs:
+        """
+        return self._geertsma_ts_custom(
+            vintage_pairs, self.subsidence.eval_geertsma_rporv, "TS_RPORV"
+        )
+
+    def geertsma_ts_simple(self, vintage_pairs):
+        """
+        Calculates TS without using velocity. Fast.
+        Velocity is only used to get the surface on the velocity grid.
+
+        :param vintage_pairs:
+        """
+        return self._geertsma_ts_custom(
+            vintage_pairs, self.subsidence.evalGeertsma, "TS_SIMPLE"
+        )
+
+    def _geertsma_ts_custom(self, vintage_pairs, subsidence_func, method_name):
+        """
+        Calculates TS without using velocity. Fast.
+
+        :param vintage_pairs:
+        :param subsidence_func: specify subsidence method to be used
+        :param method_name: string represeting the subsudence func name
+        """
+
+        if len(vintage_pairs) < 1:
+            return 0, []
+
+        vintages = self._vintages_name_date(vintage_pairs)
```

Comment:
From what I can see, what we are interested in is the unique list of vintages, and then we iterate over them. If you include the change to use a dict, and just lookup on the date/name, as suggested by @oyvindeide, you could instead have the `_vintages_name_date` return a list of Vintages e.g.:

```
        Vintage = namedtuple("Vintage", "name date")
        return [Vintage(name, date) for name, date in zip(vintages_name, vintages_date)]
```

And then in `_geerstma_ts_custom`:
```
for vintage in vintages:
    self.add_survey(vintage.name, vintage.date)
    ....
    surf_displacement[vintage.name][point] = dz2 - dz1
```

Not sure if it is even necessary with the names, which are basically "S0", "S1" etc...

Reply:
It's been refactored now.

Reply:
Done as suggested by @lars-petter-hauge